### ISSUE TITLE
JavadocTagContinuationIndentation check doesn't support arbitrary indentation on &lt;pre&gt; tags

### ISSUE THREAD

**Issue by stechio** [checkstyle/checkstyle#13048]
I have read check documentation: https://checkstyle.org/config_javadoc.html#JavadocTagContinuationIndentation
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
$ javac Test.java

$ cat test-checkstyle.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name = "Checker">
  <module name="TreeWalker">
    <module name="JavadocTagContinuationIndentation"/>
  </module>
</module>

$ cat Test.java 
/**
 * Test class.
 *
 * @apiNote
 *          This is the predefined indentation applied by Eclipse formatter to api/impl taglets.
 *          <pre>
 * Since &lt;pre> blocks retain leading whitespace, in order to avoid extra-indentation from being
 * added by the javadoc renderer, its content herein has to stick to the left margin of the comment
 * area. Unfortunately, checkstyle seems not to properly evaluate the context (&lt;pre> block) of
 * this content, and reports an error:
 *   "JavadocTagContinuationIndentation: Line continuation have incorrect indentation level,
 *   expected level should be 4"</pre>
 */
public class Test {
  public static void main(String[] args) {
  }
}

$ java -jar checkstyle-10.10.0-all.jar -c test-checkstyle.xml Test.java
Starting audit...
[ERROR] Test.java:7: Line continuation have incorrect indentation level, expected level should be 4. [JavadocTagContinuationIndentation]
[ERROR] Test.java:8: Line continuation have incorrect indentation level, expected level should be 4. [JavadocTagContinuationIndentation]
. . .
[ERROR] Test.java:12: Line continuation have incorrect indentation level, expected level should be 4. [JavadocTagContinuationIndentation]
Audit done.
Checkstyle ends with 6 errors.
```
---
CURRENT BEHAVIOR
Since &lt;pre&gt; blocks retain leading whitespace, in order to avoid extra-indentation from being added by the javadoc renderer, I have to stick their contents to the left margin of the comment area (like in `Test.java` here above) — here it is the intended result:
![Test-javadoc](https://user-images.githubusercontent.com/10812966/236636387-e7dab00a-32d7-46ad-bce2-93f001b1dad0.jpg)

Unfortunately, _checkstyle seems not to properly evaluate the context (i.e., it is unable to detect such contents are in a pre-formatted block which should be excluded from indentation enforcement)_, and raises errors of non-continuous indentation.

BTW, if, conversely, I conform the indentation of the pre-formatted block to the checkstyle rule this way...
```bash
/**
 * Test class.
 *
 * @apiNote
 *          This is the predefined indentation applied by Eclipse formatter to api/impl taglets.
 *          <pre>
 *          Since &lt;pre> blocks retain leading whitespace, in order to avoid extra-indentation from being
 *          added by the javadoc renderer, its content herein has to stick to the left margin of the comment
 *          area. Unfortunately, checkstyle seems not to properly evaluate the context (&lt;pre> block) of
 *          this content, and reports an error:
 *            "JavadocTagContinuationIndentation: Line continuation have incorrect indentation level,
 *            expected level should be 4"</pre>
 */
public class Test_extraindent {
  public static void main(String[] args) {
  }
}
```
...the resulting HTML rendering is —unsurprisingly— unacceptably over-indented (the same happens when rendered directly from source code in the Javadoc view of Eclipse IDE):
![Test_extraindent-javadoc](https://user-images.githubusercontent.com/10812966/236636453-9ee2b4b5-c64b-44fb-a6e8-c75fbf3745a2.jpg)

---
EXPECTED BEHAVIOR
Because of their inherent semantics, **pre-formatted blocks should be allowed arbitrary indentation**: once checkstyle detects an opening &lt;pre&gt; tag, it should automatically disable the `JavadocTagContinuationIndentation` check until it reaches its corresponding closing &lt;/pre&gt; tag.

**Comment by romani** [comment]
Thanks a lot for detailed report.

Please share image and javadoc of real javadoc. To let us and others be aware of such style. I have never saw it.

While we having issue, do you see a way to suppress violations?

**Comment by stechio** [comment]
> Thanks a lot for detailed report.

Thank you for maintaining such useful project.

> Please share image and javadoc of real javadoc. To let us and others be aware of such style. I have never saw it.

What do you mean with "real javadoc"? The javadoc code samples included here above have been successfully processed by the standard HTML doclet of the javadoc tool of OpenJDK 17 via cli (if you or others launch it, the result should be the same):

```bash
javadoc -d javadoc -tag apiNote:a:"API Note" Test.java
```
The screenshots included here above were taken from the resulting javadoc HTML page as rendered by Firefox 112. It's all pretty standard and easily reproducible.

> While we having issue, do you see a way to suppress violations?

In the meantime I had to arrange a file-level suppression, but that's obviously annoying as all the legit indentation checks are suppressed as well.

**Comment by romani** [comment]
I meant real comment from real code to see why people do PRE in it , and it good. Link to GitHub java code is awesome.

**Comment by stechio** [comment]
> I meant real comment from real code to see why people do PRE in it , and it good. Link to GitHub java code is awesome.

As you certainly know, `<pre>` blocks are commonly used in javadoc to denote code samples (even [SuppressWithNearbyTextFilter](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressWithNearbyTextFilter.java) itself uses them this way); JDK-specific block taglets (`@apiNote`, `@implSpec`, `@implNote`, as described in [JEP draft 8068562](https://bugs.openjdk.org/browse/JDK-8068562))  are semi-official javadoc tags extensively used in JDK documentation since version 8 (see, for example [java.util.Optional.map(..)](https://github.com/AdoptOpenJDK/openjdk-jdk11/blob/19fb8f93c59dfd791f62d41f332db9e306bc1422/src/java.base/share/classes/java/util/Optional.java#L235)), and are supported by Eclipse IDE.

*Since Eclipse IDE formatter may apply indentation to the contents of javadoc block taglets*, when `<pre>` blocks appear within them *their contents need to stick to the left margin of the comment area* in order to ensure their correct representation without incorporating that extra-indentation. My fictitious samples here above are enough to reproduce such condition (you can place whatever code you like, the syntactic issue doesn't change at all).

> Javadoc violations are not suppresable by xpath. But did you try https://checkstyle.org/config_filters.html#SuppressWithNearbyTextFilter it can help to ease a bit this problem meanwhile. Let us know if you struggling to set up to work for you.

After your suggestion, I tried the following pattern, but unfortunately it didn't work (looking at [its source code](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressWithNearbyTextFilter.java), `nearbyTextPattern` seems to be evaluated one line at a time, so my multi-line regex cannot be applied):

```xml
<module name="SuppressWithNearbyTextFilter">
  <property name="nearbyTextPattern" value="(?s)&lt;pre>.*?&lt;/pre>"/>
</module>
```

Therefore, I had to fall back on this hackish solution (`SuppressWithPlainTextCommentFilter` alone was not enough, as `onCommentFormat` restores checks immediately, on its same line; `SuppressWithNearbyTextFilter` ensures that line is suppressed too):

```xml
<module name="SuppressWithPlainTextCommentFilter">
  <property name="offCommentFormat" value="&lt;pre>"/>
  <property name="onCommentFormat" value="&lt;/pre>"/>
</module>
<module name="SuppressWithNearbyTextFilter">
  <property name="nearbyTextPattern" value="&lt;/pre>"/>
</module>
```

**Comment by romani** [comment]
Conceptually I understand how PRE might be useful, I just tried to get real life example, not artificial example, of PRE usage.

**Comment by romani** [comment]
Example of javadoc with pre:
```
$ cat Test.java
/**
 * <p>Java class for errorResponse complex type.
 *
 * <p>The following schema fragment specifies the expected content contained within this class.
 *
 * <pre>
 * &lt;complexType name="errorResponse"&gt;
 *   &lt;complexContent&gt;
 *     &lt;restriction base="{http://www.w3.org/2001/XMLSchema}anyType"&gt;
 *       &lt;sequence&gt;
 *         &lt;element name="errors" minOccurs="0"&gt;
 *           &lt;complexType&gt;
 *             &lt;complexContent&gt;
 *               &lt;restriction base="{http://www.w3.org/2001/XMLSchema}anyType"&gt;
 *                 &lt;sequence&gt;
 *                   &lt;element name="error" type="{http://www.w3.org/2001/XMLSchema}string" maxOccurs="unbounded" minOccurs="0"/&gt;
 *                 &lt;/sequence&gt;
 *               &lt;/restriction&gt;
 *             &lt;/complexContent&gt;
 *           &lt;/complexType&gt;
 *         &lt;/element&gt;
 *       &lt;/sequence&gt;
 *     &lt;/restriction&gt;
 *   &lt;/complexContent&gt;
 * &lt;/complexType&gt;
 * </pre>
 *
 *
 */
public class Test {

    protected String errors;
}
```

```
/**
 * <p>Java class for errorResponse complex type.
 *
 * <p>The following schema fragment specifies the expected content contained within this class.
 *
 * <pre>
 * &lt;complexType name="errorResponse"&gt;
 *   &lt;complexContent&gt;
 *     &lt;restriction base="{http://www.w3.org/2001/XMLSchema}anyType"&gt;
 *       &lt;sequence&gt;
 *         &lt;element name="errors" minOccurs="0"&gt;
 *           &lt;complexType&gt;
 *             &lt;complexContent&gt;
 *               &lt;restriction base="{http://www.w3.org/2001/XMLSchema}anyType"&gt;
 *                 &lt;sequence&gt;
 *                   &lt;element name="error" type="{http://www.w3.org/2001/XMLSchema}string" maxOccurs="unbounded" minOccurs="0"/&gt;
 *                 &lt;/sequence&gt;
 *               &lt;/restriction&gt;
 *             &lt;/complexContent&gt;
 *           &lt;/complexType&gt;
 *         &lt;/element&gt;
 *       &lt;/sequence&gt;
 *     &lt;/restriction&gt;
 *   &lt;/complexContent&gt;
 * &lt;/complexType&gt;
 * </pre>
 *
 *
 */
public class Test {

    protected String errors;
}
```
![image](https://github.com/checkstyle/checkstyle/assets/812984/088d4b2f-6739-47bf-9df2-631a555126d8)

**Comment by romani** [comment]
how to make java doc comment to pass javadoc tool ?
```
$ cat Test.java 
/**
 * <p>Java class for errorResponse complex type.
 *
 * <p>The following schema fragment specifies the expected content contained within this class.
 *
 * @apiNote xml of that match xsd schema
 * <pre>
 * &lt;complexType name="errorResponse"&gt;
 *   &lt;complexContent&gt;
 *     &lt;restriction base="{http://www.w3.org/2001/XMLSchema}anyType"&gt;
 *       &lt;sequence&gt;
 *         &lt;element name="errors" minOccurs="0"&gt;
 *           &lt;complexType&gt;
 *             &lt;complexContent&gt;
 *               &lt;restriction base="{http://www.w3.org/2001/XMLSchema}anyType"&gt;
 *                 &lt;sequence&gt;
 *                   &lt;element name="error" type="{http://www.w3.org/2001/XMLSchema}string" maxOccurs="unbounded" minOccurs="0"/&gt;
 *                 &lt;/sequence&gt;
 *               &lt;/restriction&gt;
 *             &lt;/complexContent&gt;
 *           &lt;/complexType&gt;
 *         &lt;/element&gt;
 *       &lt;/sequence&gt;
 *     &lt;/restriction&gt;
 *   &lt;/complexContent&gt;
 * &lt;/complexType&gt;
 * </pre>
 *
 *
 */
public class Test {

    protected String errors;
}rivanov@p5510:/var/tmp$ javadoc Test.java
Loading source file Test.java...
Constructing Javadoc information...
Standard Doclet version 11.0.14.1
Building tree for all the packages and classes...
Test.java:6: error: unknown tag: apiNote
 * @apiNote xml of that match xsd schema
   ^
Generating ./Test.html...
Generating ./package-summary.html...
Generating ./package-tree.html...
Generating ./constant-values.html...
Building index for all the packages and classes...
Generating ./overview-tree.html...
Generating ./index-all.html...
Building index for all classes...
Generating ./allclasses-index.html...
Generating ./allpackages-index.html...
Generating ./deprecated-list.html...
Building index for all classes...
Generating ./allclasses.html...
Generating ./allclasses.html...
Generating ./index.html...
Generating ./help-doc.html...
1 error

``` 

@stechio , do you know ?

**Comment by stechio** [comment]
You can generate javadoc with JDK-specific block taglets as reported in [my previous reply](https://github.com/checkstyle/checkstyle/issues/13048#issuecomment-1537482953)...
```bash
javadoc -d javadoc -tag apiNote:a:"API Note" Test.java
```
...BUT the problem is NOT those special tags: the problem is that **pre-formatted blocks currently are not allowed arbitrary indentation** (no matter the context where they are placed — special tags or whatever else). Such limitation emerges if the block containing a `<pre>` is formatted with extra-indentation, such as when Eclipse IDE automatically formats those special tags, this way:

```java
/**
 * <p>Java class for errorResponse complex type.
 *
 * <p>The following schema fragment specifies the expected content contained within this class.
 *
 * @apiNote
 *          xml of that match xsd schema
 *          <pre>
 * &lt;complexType name="errorResponse"&gt;
 *   &lt;complexContent&gt;
 *     &lt;restriction base="{http://www.w3.org/2001/XMLSchema}anyType"&gt;
 *       &lt;sequence&gt;
 *         &lt;element name="errors" minOccurs="0"&gt;
 *           &lt;complexType&gt;
 *             &lt;complexContent&gt;
 *               &lt;restriction base="{http://www.w3.org/2001/XMLSchema}anyType"&gt;
 *                 &lt;sequence&gt;
 *                   &lt;element name="error" type="{http://www.w3.org/2001/XMLSchema}string" maxOccurs="unbounded" minOccurs="0"/&gt;
 *                 &lt;/sequence&gt;
 *               &lt;/restriction&gt;
 *             &lt;/complexContent&gt;
 *           &lt;/complexType&gt;
 *         &lt;/element&gt;
 *       &lt;/sequence&gt;
 *     &lt;/restriction&gt;
 *   &lt;/complexContent&gt;
 * &lt;/complexType&gt;</pre>
 */
public class Test {
```

**Comment by romani** [comment]
I just want share here how it looks like html to share real usage of it.  The more realistic and visual the problem definition the more chance it will be fixed by somebody.

**Comment by romani** [comment]
@stechio , please share details ^^
